### ISSUE TITLE
Null Subsitute causing MappingConfig to throw Argument types do not match 

### ISSUE THREAD

**Issue by Zhobday** [AutoMapper/AutoMapper#1589]
```
config.CreateMap<Citation, CitationViewModel>().
                ForMember(x => x.Source, y => y.NullSubstitute(new ReportViewModel()));
```

When this code is in the the mapping config doesnt compile in V.5 is there a new method or way of dealing with mapping null values?

**Comment by lbargaoanu** [comment]
Try the MyGet build. The null substitute should have the type of the source for that property.

**Comment by Zhobday** [comment]
I'm in VS and they do have the same types

**Comment by lbargaoanu** [comment]
If you have a failing test with 5.1.0-alpha-01196 from MyGet I want to see it. Otherwise the issue can be closed.

**Comment by koljada** [comment]
I also have this bug with AutoMapper 5.0.2 nuget version.

**Comment by lbargaoanu** [comment]
@koljada That's only interesting if the same thing happens with the MyGet build.

**Comment by jbogard** [comment]
@Zhobday can you include some source/destination types? I can't easily reproduce on a single line of config.

**Comment by tillig** [comment]
Just ran into this myself. Here's a repro using 5.0.2:

```csharp
public class Fixture
{
  [Fact]
  public void Convert_Dictionary_EmptyDestination()
  {
    var source = new Source();
    var mapper = CreateMapper();
    var dest = mapper.Map<Destination>(source);
    Assert.Equal("null!", dest.Model.Data);
  }

  private static IMapper CreateMapper()
  {
    return new MapperConfiguration(cfg =>
    {
      cfg.CreateMap<Source, Destination>()
        .ForMember(dest => dest.Model, opt => opt.NullSubstitute(new DestinationModel { Data = "null!" }));
      cfg.CreateMap<SourceModel, DestinationModel>();
    }).CreateMapper();
  }

  private class Source
  {
    public SourceModel Model { get; set; }
  }

  private class SourceModel
  {
    public string Data { get; set; }
  }

  private class Destination
  {
    public DestinationModel Model { get; set; }
  }

  private class DestinationModel
  {
    public string Data { get; set; }
  }
}
```

In 5.0.2 you get the exception: `System.ArgumentException : Argument types do not match`

```
at System.Linq.Expressions.Expression.ValidateCoalesceArgTypes(Type left, Type right)
   at System.Linq.Expressions.Expression.Coalesce(Expression left, Expression right, LambdaExpression conversion)
   at System.Linq.Expressions.Expression.MakeBinary(ExpressionType binaryType, Expression left, Expression right, Boolean liftToNull, MethodInfo method, LambdaExpression conversion)
   at System.Linq.Expressions.Expression.MakeBinary(ExpressionType binaryType, Expression left, Expression right)
   at AutoMapper.Execution.TypeMapPlanBuilder.BuildValueResolverFunc(PropertyMap propertyMap, TypeMapRegistry typeMapRegistry, ParameterExpression srcParam, ParameterExpression destParam, Expression destValueExpr, ParameterExpression ctxtParam)
   at AutoMapper.Execution.TypeMapPlanBuilder.CreatePropertyMapFunc(PropertyMap propertyMap, IConfigurationProvider configurationProvider, TypeMapRegistry typeMapRegistry, ParameterExpression srcParam, ParameterExpression destParam, ParameterExpression ctxtParam)
   at AutoMapper.Execution.TypeMapPlanBuilder.TryPropertyMap(PropertyMap pm, IConfigurationProvider configurationProvider, TypeMapRegistry registry, ParameterExpression srcParam, ParameterExpression destParam, ParameterExpression ctxtParam)
   at AutoMapper.Execution.TypeMapPlanBuilder.CreateAssignmentFunc(TypeMap typeMap, IConfigurationProvider configurationProvider, TypeMapRegistry registry, ParameterExpression srcParam, ParameterExpression destParam, ParameterExpression ctxtParam, Expression destinationFunc, ParameterExpression destination, Boolean constructorMapping)
   at AutoMapper.Execution.TypeMapPlanBuilder.BuildMapperFunc(TypeMap typeMap, IConfigurationProvider configurationProvider, TypeMapRegistry typeMapRegistry)
   at AutoMapper.TypeMap.Seal(TypeMapRegistry typeMapRegistry, IConfigurationProvider configurationProvider)
   at AutoMapper.MapperConfiguration.Seal(IConfiguration configuration)
   at AutoMapper.MapperConfiguration..ctor(MapperConfigurationExpression configurationExpression, IEnumerable`1 mappers)
   at AutoMapper.MapperConfiguration..ctor(Action`1 configure, IEnumerable`1 mappers)
   at AutoMapper.MapperConfiguration..ctor(Action`1 configure)
   at Library.Test.Fixture.CreateMapper() in C:\...\Class1.cs:line 23
```

What's interesting is that this test also fails under 5.1.1, but with a different exception.

`System.InvalidOperationException : No coercion operator is defined between types 'Library.Test.Fixture+DestinationModel' and 'Library.Test.Fixture+SourceModel'.`

```
at System.Linq.Expressions.Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)
   at System.Linq.Expressions.Expression.Convert(Expression expression, Type type, MethodInfo method)
   at AutoMapper.ExpressionExtensions.ToType(Expression expression, Type type)
   at AutoMapper.Execution.TypeMapPlanBuilder.BuildValueResolverFunc(PropertyMap propertyMap, Expression destValueExpr)
   at AutoMapper.Execution.TypeMapPlanBuilder.CreatePropertyMapFunc(PropertyMap propertyMap)
   at AutoMapper.Execution.TypeMapPlanBuilder.TryPropertyMap(PropertyMap propertyMap)
   at AutoMapper.Execution.TypeMapPlanBuilder.CreateAssignmentFunc(Expression destinationFunc, Boolean constructorMapping)
   at AutoMapper.Execution.TypeMapPlanBuilder.CreateMapperLambda()
   at AutoMapper.TypeMap.Seal(TypeMapRegistry typeMapRegistry, IConfigurationProvider configurationProvider)
   at AutoMapper.MapperConfiguration.Seal(IConfiguration configuration)
   at AutoMapper.MapperConfiguration..ctor(MapperConfigurationExpression configurationExpression, IEnumerable`1 mappers)
   at AutoMapper.MapperConfiguration..ctor(Action`1 configure)
   at Library.Test.Fixture.CreateMapper() in C:\...\Class1.cs:line 23
```

Basically, it 5.1.1 wants the null substitute to be of the _source_ type instead of the _destination_ type, which is a breaking change.

You can fix it by switching the mapper configuration:

```csharp
private static IMapper CreateMapper()
{
  return new MapperConfiguration(cfg =>
  {
    cfg.CreateMap<Source, Destination>()
       .ForMember(dest => dest.Model, opt => opt.NullSubstitute(new SourceModel { Data = "null!" }));
    cfg.CreateMap<SourceModel, DestinationModel>();
  }).CreateMapper();
}
```

In all of the 5.x series, I've found the only real workaround is to add an `AfterMap` handler to set the null value on the destination if it hasn't already been set.

**Comment by jbogard** [comment]
I went about trying to fix this, but it was pretty tough. When you move to letting the destination type dictate the null substitute, it messes with the overall mapping flow.

With the new way, the mapping flow is:

- Resolve value
- Map resolved value to destination value
- Assign mapped value to destination member

We wrap the "Resolve" step in the null check, then map. If we map, then check for nulls, we have the problem where the resolved value could be a value type, so we'd have to check in multiple places for the null substitute. It just got really ugly.

So, this is the new expected behavior, I'll update the docs accordingly.
